### PR TITLE
feat(acir): Handle dynamic array operations for nested slices

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Works towards getting slices up-to-par with arrays with their support. I decided to separate out the ACIR work into a separate PR as full support is predicated upon fixing flattening in SSA to support merging of nested slices.

The majority of the SSA changes were handled in this preceding PR (https://github.com/noir-lang/noir/pull/3258). They were necessary to get basic dynamic array operations on nested slices working at all. The focus in this PR are the changes outlined below in the Summary section.

## Summary\*

There are two major changes in this PR:
1. The way the element type sizes array is constructed is now based upon an array's internal element types but rather its SSA values. This allows us to fetch to convert the SSA array index to the flattened index we expect when working with memory operations, as mem ops in the ACVM work with a flat memory structure.
2. We construct a map of a slices to their own length in the SSA, as well as each of its internal slices sizes. This is necessary to maintain the internal type structure as we move through ACIR gen. We only construct this slice map when we run into a `Value::Array`. Otherwise it is assumed that the `array_get` and `array_set` operations associate their results with the correct slice sizes map. This allows us to leave the logic of initializing the element type sizes array when we have an `AcirValue::Array` or `AcirValue::DynamicArray` largely the same. In order to have matches internal slice sizes an SSA pass was made to fill internal slices with dummies so they match in size (https://github.com/noir-lang/noir/pull/3258).

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
@vezenovm is this ready for review? I was about to review it but saw you recently marked it as a draft.

**Comment by vezenovm** [comment]
@jfecher I marked it back as a draft as I have found some issues while working on flattening nested slices. I would hold off on a review for now.

**Comment by jfecher** [comment]
Is this PR succeeded by the new slice PR?

**Comment by vezenovm** [comment]
> Is this PR succeeded by the new slice PR?

The new PR is branched off of this one. I made a separate PR to have some more separation as this PR is mainly the initial ACIR gen changes while this [followup PR](https://github.com/noir-lang/noir/pull/3258) is mostly the SSA pass. The new pass has revealed some of its own ACIR gen updates but this one has the majority of ACIR gen changes.

**Comment by TomAFrench** [comment]
@vezenovm mind if I push a merge commit?

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1292]: The goal was that in the case of a true predicate we use the normal var index, while in the case of a false predicate we just set the var index to zero as otherwise we could have memory out of bounds errors. As I am just reading from zero in the false though, I see what you mean about them being equivalent. Just want to test it to make sure before committing

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:928]:
```suggestion
                let current_size = slice_sizes.pop_front();
```

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1011]:
```suggestion
            let has_internal_slices =
                element_types.as_ref().iter().any(|typ| typ.contains_slice_element());
```

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1156]:
```suggestion
                                    let type_sizes_array_len = 
                                        self.internal_mem_block_lengths.get(&inner_elem_type_sizes).copied().ok_or_else(||
                                        InternalError::General {
                                            message: format!("Array {array_id}'s inner element type sizes array does not have a tracked length"),
                                            call_stack: self.acir_context.get_call_stack(),
                                        }
                                    )?;
```

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1258]: We can remove two layers of nesting here with an early return.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1292]: Ah ok, that makes sense. We should be able to simplify this all down to `let pred_var_index =
            self.acir_context.mul_var(var_index, self.current_side_effects_enabled_var)?;` in that case.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:753]:
```suggestion
                    self.slice_sizes.get(&store).cloned().ok_or_else(|| {
                        InternalError::UnExpected {
                            expected: "Store value should have slice sizes computed".to_owned(),
                            found: "Missing key in slice sizes map".to_owned(),
                            call_stack: self.acir_context.get_call_stack(),
                        }
                    })?
```

**Review by TomAFrench — COMMENTED** [review]
Mostly code style suggestions to reduce branching/nesting.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:928]: This is switched to a slice in the followup PR https://github.com/noir-lang/noir/pull/3258 so it looks a bit different now

**Comment by vezenovm** [comment]
> Mostly code style suggestions to reduce branching/nesting.

Will apply these changes once we are merged with https://github.com/noir-lang/noir/pull/3258

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1290]: Yup you are correct, done.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: The diff got out of date so I applied this suggestion manually.

Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: The diff got out of date so I applied this suggestion manually.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: done

**Comment by vezenovm** [comment]
This PR is ready for review. The majority of the changes are from #3258 so you can just foucs on the changes laid out in the description for ACIR gen.

**Comment by jfecher** [comment]
If the majority of changes are from #3258 then I'm not sure why the diff is so large when those should be excluded

**Comment by vezenovm** [comment]
> If the majority of changes are from #3258 then I'm not sure why the diff is so large when those should be excluded

I made the error of merging that PR into this one, when we should have merged this one into master then #3258 into master.

**Comment by jfecher** [comment]
> I made the error of merging that PR into this one, when we should have merged this one into master then https://github.com/noir-lang/noir/pull/3258 into master.

Can you re-open #3258 then with master as a merge target? It is too difficult to review this >1000 line PR when most changes are from a different PR

**Comment by vezenovm** [comment]
> Can you re-open #3258 then with master as a merge target? It is too difficult to review this >1000 line PR when most changes are from a different PR

#3258 has been reopened here (https://github.com/noir-lang/noir/pull/3410)

File: tooling/nargo_cli/tests/execution_success/nested_array_dynamic/src/main.nr
Comment (TomAFrench) [inline tooling/nargo_cli/tests/execution_success/nested_array_dynamic/src/main.nr:33]:
```suggestion
    assert(x[y].a == 50);
```

What's the reasoning behind changing this?

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:750]:
```suggestion
                    self.slice_sizes.get(&store).copied().ok_or_else(|| {
```

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1247]:
```suggestion
        if let Some(parent_array) = parent_array {
            let sizes_list =
                self.slice_sizes.get_mut(&parent_array).expect("ICE: expected size list");
            sizes_list.push(true_len);
            for value in array {
                self.compute_slice_sizes(*value, parent_array, dfg);
            }
        } else {
            // This means the current_array_id is the parent array
            self.slice_sizes.insert(current_array_id, vec![true_len]);
            for value in array {
                self.compute_slice_sizes(*value, Some(current_array_id), dfg);
            }
        }
```

Personal taste but I think it makes sense to merge these two if-elses.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:875]:
```suggestion
            let slice_sizes = self
                .slice_sizes
                .get(&array_id)
                .expect("ICE: Array with slices should have associated slice sizes");

            let slice_sizes = slice_sizes[1..].to_vec();
```

Nit but we can avoid the shifting from `.remove` here.

File: tooling/nargo_cli/tests/execution_success/nested_array_dynamic/src/main.nr
Comment (vezenovm) [reply tooling/nargo_cli/tests/execution_success/nested_array_dynamic/src/main.nr:33]: I was thinking to do this change so that we guarantee that the dynamic index I change inside the if statement is actualy `3`. If `x[y].a` actually wrote to some other index asserting on `y` later would still hold.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (TomAFrench) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:875]: Also I forget, we're slicing off the first element as it's holding the length of the array, correct? Could we sprinkle some comments in places where we're adding/removing this element and it's not obvious from how we're using it?

**Review by TomAFrench — APPROVED** [review]
Some final nits, looks good to me.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: I'm not sure what the time difference is with a clone vs. a shift but I accepted the suggestion.

Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: I added a couple comments for this as well.

Comment (jfecher) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: > I'm not sure what the time difference is with a clone vs. a shift but I accepted the suggestion.

In the first case we're cloning then removing, both O(n).
In the second case we're creating a slice from the array O(1), then cloning O(n). So slicing first instead of removing after removes a O(n) operation.

File: compiler/noirc_evaluator/src/ssa/ir/types.rs
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/ir/types.rs:92]: Can you remove the catch-all here? I'd like this to error if we ever add/change types in the future.
It also seems odd to me to panic instead of handling the additional cases when the method is named so generally. Can you extend this to cover the missing cases as well instead of panicking?

Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/ir/types.rs]: I removed the catch-all. I was originally unsure if I should handle references differently but I believe it is correct to just return false in this instance as this method is used for simply determining whether we should perform a read differently when we are returning a slice from an array_get or storing a slice in an array_set. A reference to a slice shouldn't be treated as a slice type itself.

**Review by jfecher — APPROVED** [review]
:+1:

**Pull request merged**
